\documentclass[11pt]{article}

\usepackage{comment} % enables the use of multi-line comments (\ifx \fi) 
\usepackage[a4paper,margin=1cm]{geometry}
\usepackage[utf8]{inputenc}
\usepackage[ngerman]{isodate}
\usepackage{gensymb}
\usepackage{graphicx}
\usepackage{booktabs}% http://ctan.org/pkg/booktabs
\usepackage{tabularx}
\usepackage{ltablex} % Longtables with tabularx
\usepackage[x11names]{xcolor}
\usepackage{amsmath}
\usepackage{amssymb}
\usepackage{amsthm}
\usepackage{array}
\usepackage{wrapfig}
\usepackage{subcaption}
\usepackage{csquotes}
\usepackage{pdflscape}
\usepackage{geometry}
\usepackage{multicol}
\usepackage{bm}
\usepackage{enumitem}
\usepackage{hyperref}
\usepackage{mdframed}
\usepackage{scalerel}
\usepackage{stackengine}
\usepackage{mathtools}
\usepackage{pdfpages}

% Code highlighting
\usepackage{minted}
\surroundwithmdframed{minted}

% Be able to caption equations and float them in place
\usepackage{float}

\newmdtheoremenv{theorem}{Theorem}

\theoremstyle{definition}
\newmdtheoremenv{definition}{Definition}[section]


\geometry{a4paper, margin=2.4cm}

\newcommand\equalhat{\mathrel{\stackon[1.5pt]{=}{\stretchto{\scalerel*[\widthof{=}]{\wedge}{\rule{1ex}{3ex}}}{0.5ex}}}}
\newcommand\defeq{\mathrel{\overset{\makebox[0pt]{\mbox{\normalfont\tiny def}}}{=}}}
\newcolumntype{C}{>{\centering\arraybackslash}X}

\newcommand*\samplemean[1]{\overline{#1}}
\newcommand*\ev[1]{\mathrel{\mathbb{E}\left[#1\right]}}
\newcommand*\R{\mathbb{R}}
\newcommand*\Z{\mathbb{Z}}
\newcommand*\N[1]{\mathcal{N}\left(#1\right)}
\newcommand*\Likelihood{\mathcal{L}}
\newcommand*\diff{\mathop{}\!\mathrm{d}}
\newcommand*\Diff[1]{\mathop{}\!\mathrm{d^#1}}
\newcommand*\Exp[1]{\mathop{\text{Exp}}\left(#1\right)}
\newcommand*\Cov[1]{\mathop{\text{Cov}}\left(#1\right)}
\newcommand*\Cor[1]{\mathop{\text{Cor}}\left(#1\right)}
\newcommand*\Var[1]{\mathop{\text{Var}}\left(#1\right)}

\newcommand{\incond}{\rotatebox[origin=c]{90}{$\vDash$}}
\newcommand{\dcond}{\rotatebox[origin=c]{90}{$\nvDash$}}

\DeclarePairedDelimiter\abs{\lvert}{\rvert}
\DeclarePairedDelimiter\norm{\lVert}{\rVert}

\setcounter{tocdepth}{3}
\setcounter{secnumdepth}{3}

\graphicspath{{./img/}}

\begin{document}
	
\title{Machine Learning FS20}
\author{Pascal Baumann\\pascal.baumann@stud.hslu.ch}
\maketitle



For errors or improvement raise an issue or make a pull request on the \href{https://github.com/KilnOfTheSecondFlame/mse_summaries}{github repository}.

\tableofcontents
\newpage



\section{Introduction}

\subsection{Inductive Learning}

Inductive learning has the goal to \textbf{discover general concepts} from a \textbf{limited set of examples}, a model using inductive reasoning obtains new general knowledge from specific information. Throughout the learning process it is not truth preserving, new information can invalidate the current model, and works heuristically. The assumption therefore is that a model fitted to a sufficiently large learning set will be able to generalise to unseen data.

\begin{landscape}
	\subsubsection{Machine Learning Flowchart}
	\begin{center}
		\includegraphics[height=0.9\textheight, keepaspectratio]{scikit-learn_algorithm_cheatsheet}
	\end{center}
\end{landscape}

\subsection{Inductive Supervised Learning}
The data usually has a semi-formal representation in attribute-label pairs $(\textbf{x}, y)$. The labels encode concepts or classes.
Approximate the mapping function from example $x$ to label $y$ with a hypothesis $h(x) = \hat{y} \approx f(x)$ so that it generalises well.

\subsection{Learning as Search}
\begin{center}
	\includegraphics[width=0.7\linewidth]{img/hypothesis_space}
\end{center}

The hypothesis space $\mathcal{H}$ contains all possible hypotheses that can be built with the chosen representation, learning can thus be understood as a search for the global minima in this space $\mathcal{H}$.

\noindent
Formally this goal is expressed as follows: Find the hypothesis $h^* (x,\theta) = \hat{y}$ that best fits the training data, according to a loss function $L(h(x,\theta))$ by searching the hypothesis space $\mathcal{H} = \left\{ h(x,\theta) \middle| \theta \in P\right\}$.

\subsection{Inductive Bias}
\begin{theorem}
	A learner that makes \textbf{no a-priori assumptions} regarding the identity of the target concept has \textbf{no rational basis for classifying} any unseen instances. - Mitchell
\end{theorem}
All learning algorithms have some preformed hypothesis that is helpful to look for. It is beneficial to choose algorithms whose implicit hypothesis fits the data.

No free lunch theorem regarding the general equivalence of learners states that if all functions $f$ are equally likely, the probability of observing an arbitrary sequence of cost values during training does not depend on the learning algorithm $\mathcal{L}$.

The inductive bias of a learning algorithm $\mathcal{L}$ for instances in $X$ are any \textbf{minimal set of assertions} $B$ that, together with $\mathcal{L}$ and the training set $D$ \textbf{allows for deductively inferring} the $y'$ for a new $x\in X$. Make all assumptions \textbf{explicit} in $B$ such that $\forall x' \in X: \left(B, \mathcal{L}, D, x' \right) \Rightarrow y'$ is provable.

Machine Learning depends on the intelligent choice of the class $\mathcal{H}$ where $\mathcal{L}$ optimises the parameters. Machine Learning algorithms can be categorised by the strength of their inductive bias.

\subsection{Inductive Unsupervised Learning}
\begin{minipage}{0.6\linewidth}
	The usual task is Clustering, where the data $D$ are described by feature vectors without any labels, interesting structures and relationships in the data are then searched. The data naturally fall into $K$ groups, where the number of classes are not determined at the start but found through the learning scheme. This presents the challenge of searching by similarity in \textbf{distance} or \textbf{density}, which is another inductive bias to be considered, and by the choice of \textbf{parameters}.
	
\end{minipage}
\begin{minipage}{0.4\linewidth}
	\begin{center}
		\includegraphics[width=0.6\linewidth]{img/clustering_example}
	\end{center}
\end{minipage}

\subsection{Learnability}
Any target function $f$ over an instance $X$ is learnable given an expressive enough deterministic hypothesis space $\mathcal{H}$, a large enough training set $D$, and stationarity of the distribution $X$, that is instances in $D_{\text{train}}$ and $D_{\text{test}}$ are independently, identically distributed.

Better questions are what size of \textbf{$D_{\text{train}}$} are large enough and given a large enough training set, how well does the \textbf{training error predict generalisability}. These questions are in the domain of \emph{computational learning theory}.

\subsection{Probably Approximate Correct Learning and Vapnik-Chervonenkis Complexity}
Sample complexity bounds using $\abs{\mathcal{H}}$ usually do substantially overestimate, thus the assumption that $\bm{f} \in \mathcal{H}$ is unrealistic.

Measuring Vapnik-Chervonenkis (VC) dimension for infinite $\mathcal{H}$:
$h\in\mathcal{H}$ is \textbf{shattering a set of instances} $S\in X$ \textbf{iif} $h$ can partition $S$ in any way possible.\\
$VC(\mathcal{H}) := \abs{\{S\in X | S \text{ is the largest subset of } X \text{ shattered by any } h\in\mathcal{H}\}}$. $VC(\mathcal{H})$ can be used as an alternative measure of $\abs{\mathcal{H}}$ to compute sample complexity.

\begin{figure}[H]
	\centering
	\includegraphics[keepaspectratio,width=0.9\linewidth]{VC_shattering}
	\caption{A two-dimensional linear classifier can shatter three points, $VC(\text{two-dimensional straight lines}=3$ but $\abs{\mathcal{H}} = \infty$}
\end{figure}

\section{Formulating Learning Problems}
\subsection{Designing a Learning Solution}
\begin{center}
	\includegraphics[width=\linewidth,keepaspectratio]{designing_learning_solution.pdf}
\end{center}

Designing a problem that is easiest for the algorithm to solve is the best thing one could do while formulating the problem. The first thing to do while first analysing data is to plot it.

\subsection{Machine Learning Development Process - Exploration and Experimentation}
There is a necessity to have a distinct conceptual approach to model data. Focus on \textbf{systematic experimentation} and \textbf{rigorous evaluation}, even automatised if needed. This is best implemented by a \textbf{pipeline of scripts} similar to the UNIX command line approach. Data exploration and rapid prototyping is key in the beginning of this process.

\begin{center}
	\includegraphics[width=0.6\linewidth]{exploration_evaluation_pipeline}
\end{center}

\section{Machine Learning Guiding Principles}

\subsection{Cost Function J}
Choose $\theta_0, \theta_1$ in such a way that $h(x)$ is close to $y$ for the training examples $(x,y)$. Let a function $J$ \textbf{number the cost of errors made} for the specific set of parameters.

\begin{equation*}
	J_{MSE}(\theta_0, \theta_1) = \frac{1}{2N} \sum_{i=1}^{N} \left(h(x_i, \vec{\theta}) - y_i\right)^2
\end{equation*}

The objective is to minimise $J$ with regards to $\theta_0, \theta_1$

\subsection{Optimisation by Gradient Descent}
Algorithm
\begin{itemize}
	\item Start with a random $\theta_0, \theta_1$
	\item Keep changing $\theta_0, \theta_1$ in a way to reduce $J(\theta_0, \theta_1)$
	\item End at a local minimum
\end{itemize}

\begin{align*}
	\text{\texttt{tmp\_0}} &:= \theta_0 - \alpha \cdot \frac{\partial J}{\partial \theta_0} J(\theta_0,\theta_1)\\
	\text{\texttt{tmp\_1}} &:= \theta_1 - \alpha \cdot \frac{\partial J}{\partial \theta_1} J(\theta_0,\theta_1)\\
	\theta_0 &:= \text{\texttt{tmp\_0}}\\
	\theta_1 &:= \text{\texttt{tmp\_1}}\\
\end{align*}

$\alpha$ is called the learning rate, and is a data-dependent hyper-parameter of the algorithm.

\section{Model Assessment and Selection}

\subsection{Data Handling for Model Evaluation}
Search for methods that help to learn and evaluate algorithms based on \emph{limited data} and help to \emph{deduce the true error} from the training error.

\begin{itemize}[leftmargin=*, labelindent=3cm, labelsep=1cm]
	\item[Model Assessment] \textbf{Evaluate} a model's \textbf{performance}
	\item[Model Selection] \textbf{Select} among competing models the one with the \textbf{proper level of flexibility}
\end{itemize}

\begin{center}
	\includegraphics[width=0.9\linewidth]{img/model_flexibility}
\end{center}

\subsection{Optimal Usage of a Small Dataset}
Split into Training, Validation and Test set. Do k-fold Cross Validation on Training and Validation set, final test on Test set. Train k times on $(l - 1)$ folds, validate on the remaining one (until each fold was used for validation once), then average the error.

\subsection{Observable and Unobservable Errors}
True error $E_D$ is the probability that $h$ will misclassify a random instance from the \textbf{complete domain} $\bm{E}$ and is unobservable.

The empirical test error $E_{emp}$ is the proportion of \textbf{sample $S\in D$} misclassified by $h$ is en estimate for the true error and gets better with more error.

\begin{itemize}
	\item Assumption that training and test data are representative of underlying distribution of $D$
	\item $S$ and $h$ are usually not chosen independently, that means that the test error is \textbf{optimistically biased}
	\item The test error usually varies for different $S\in D$, that means it has a higher variance than the true error
\end{itemize}

\subsection{Error Sources}

The chosen hypothesis $\mathcal{H}$ is the \textbf{best hypothesis at a distance of the true function}. Different chosen samples $X$ give \textbf{different information}.

\noindent
Error decomposition:
\begin{equation*}
	E_{MSE} = \underbrace{\text{systematic error}}_{\text{bias}} + \underbrace{\text{dependence on specific sample}}_{\text{variance}}  + \underbrace{\text{random nature of process}}_{\text{irreducible noise or Bayes rate}}
\end{equation*}

\subsection{ROC Curve}
First used in signal detection to show trade-off between \textbf{hit rate} and \textbf{false alarm rate} over noisy channel

\begin{center}
	\includegraphics[width=0.7\linewidth]{img/ROC_curve}
\end{center}

\noindent
\textbf{Interpretation}
\begin{itemize}[leftmargin=*, labelindent=3cm, labelsep=1cm]
	\item[Straight Line] Indicates a random process
	\item[Jagged Curve] Created with one set of test data
	\item[Smooth Curve] Created using averages
\end{itemize}

\subsection{Error Measures}
\begin{tabularx}{\linewidth}{lX}
	Mean Square Error & $ E_{MSE} = \frac{1}{F}\sum_{i=1}^{N} (\widehat{y}_i - y_i)^2 $\\
	Root Mean-Squared Error & $E_{RMSE} = \sqrt{\frac{1}{F}\sum_{i=1}^{N} (\widehat{y}_i - y_i)^2 } $\\
	Mean Absolute Error & $ E_{MAE} = \frac{1}{F}\sum_{i=1}^{N} \abs{\widehat{y}_i - y_i}$\\
	(less sensitive to outliers)
\end{tabularx}

\subsection{Evaluating Clustering Methods}
\textbf{Without Labels}:\\
Use the \textbf{silhouette coefficient}, which is a measure for cluster validity or consistency, do a visual inspection of dendrograms and a visual comparison of the dimension reduction o the feature vectors.

\vspace{1em}
\noindent
\textbf{With ground truth available}:\\
Check \textbf{purity}, \textbf{rand index} and \textbf{misclassification rate}.

\subsection{Selection Among Competing Models}
On what basis can one \emph{algorithm be favoured over another} and how probable is it that the chosen method is \emph{truly significantly better}.

\subsection{Maximum Likelihood and Ockham's Razor}
Given competing $h_i \in \mathcal{H}_j$ \textbf{maximum likelihood parameters} can be found by \textbf{calculating the likelihood} $p(X|h_i)$ and \textbf{selecting the best} $\hat{h} = \underset{h_i}{\max} p(X|h_i)$.

The goal is to find a compromise between model complexity and accuracy on the validation data. Ockham's razor, an axiom in machine learning, states that "given two models with the \textbf{same empirical error}, the simpler one should be preferred because \textbf{simplicity is desirable in itself}". The reasoning is that for a simple hypothesis the \textbf{probability of it having unnecessary conditions is reduced}.

\subsection{Determining True Best Classifier}
In practice tenfold cross validation is often good enough and it does not matter which classifier is really better. Otherwise the \textbf{Student's t-test} can be used to compare two samples. Generally, the difference between $\mu_{\mathcal{L}_A}$ and $\mu_{\mathcal{L}_A}$ of the obtained cross-validation error estimates follows a Student's distribution with $m-1$ degrees of freedom.

If the number of features $p$ is large in comparison to the number of instances $N$ use \textbf{boosting} or \textbf{SVM}. If the data set is \emph{severely imbalanced}, \textbf{non-standard loss functions} that take class distribution in account or \textbf{Bayesian methods} and appropriate prior probabilities should be considered.

\section{Support Vector Machines}

Each observation is a vector of values ($p$-dimensional) for which the SVM \textbf{constructs a hyperplane} to separate class members. There are different support vectors for different tasks
\begin{enumerate}
	\item Maximal margin \textbf{hyperplane} classifier, for linearly separable data
	\item Support vector \textbf{classifier}, for almost linearly separable data
	\item Support vector \textbf{machine}, for non-linearly separable data
\end{enumerate}

\subsection{Maximal Margin Classifier}
\begin{minipage}{0.6\linewidth}
	Chooses the hyperplane maximising the distance from the hyperplane to the closest training point, denoted the support vectors, and can be represented as a linear combination of only a few training points.
\end{minipage}
\begin{minipage}{0.4\linewidth}
	\centering
	\includegraphics[keepaspectratio,width=0.8\linewidth]{img/maximum_margin_classifier.png}
\end{minipage}

\begin{definition}
	Definition of a hyperplane
	\begin{equation*}
		\{ \vec{x}: \beta_0 + \beta_1 x_1 + \beta_2 x_2 + \dots + \beta_p x_p = 0 \}
	\end{equation*}
\end{definition}
Separating hyperrplanes for classes encoded $\pm 1$
\begin{align*}
	\beta_0 + \beta_1 x_{i1} + \beta_2 x_{i2} + \dots + \beta_p x_{ip} &> 0 \qquad \text{ if } y_i = 1\\
	\beta_0 + \beta_1 x_{i1} + \beta_2 x_{i2} + \dots + \beta_p x_{ip} &< 0 \qquad \text{ if } y_i = -1\\
	\intertext{Equivalently}
	y_i\cdot (\beta_0 + \beta_1 x_{i1} + \beta_2 x_{i2} + \dots + \beta_p x_{ip}) &> 0
\end{align*}
Evaluating the hyperplane formula for a point $x$ gives its distance to the hyperplane, if $\bm{\beta}$ is a normal vector.

\subsection{Formulation of Optimisation Problem}
Intuitive Optimisation
\begin{itemize}[noitemsep]
	\item $\underset{\beta_0,\beta_1,\dots,\beta_p}{\text{maximise}} M$
	\begin{itemize}
		\item subject to $\sum_{j=1}^{p}\beta_j^2 = 1$
		\item and $y_i\cdot (\beta_0 + \beta_1 x_{i1} + \beta_2 x_{i2} + \dots + \beta_p x_{ip}) \geq M\qquad\forall i = 1..N$
	\end{itemize}
\end{itemize}
This can be reformulated using {\color{DodgerBlue2} Lagrange multipliers}
\begin{itemize}[noitemsep]
	\item Find $L_D = \sum_{i=1}^{N} \alpha_i - \frac{1}{2} \sum_{i=1}^{N}\left( \sum_{k=1}^{N} \alpha_i \alpha_k \bm{x}_i^T \bm{x}_k \right) $
		\begin{itemize}
		\item subject to $ \alpha_i \geq 0 $ and $\sum_{i=1}^{N} \alpha_i y_i = 0$
		\item Once the alphas are computed $\beta = \sum_{i=1}^{N} \alpha_i y_i \bm{x}_i$
	\end{itemize}
\end{itemize}

\subsection{Support Vector Classifier}
Based on the assumption that the data is not perfectly linearly separable any more, the idea is to introduce soft margins that allow for \emph{some} misclassifications. The number of misclassifications is controlled by a {\color{DodgerBlue2} penalty factor} $C$.

\begin{center}
	\includegraphics[width=0.7\linewidth]{img/support_vector_classifier}
\end{center}

\subsubsection{Support Vector Classifier Optimisation}
\begin{minipage}{0.6\linewidth}
Intuitive Optimisation
\begin{itemize}[noitemsep,nosep]
	\item $\underset{\beta_0,\beta_1,\dots,\beta_p}{\text{maximise}} M$
	\begin{itemize}
		\item subject to $\sum_{j=1}^{p}\beta_j^2 = 1$
		\item and $y_i\cdot (\beta_0 + \beta_1 x_{i1} + \beta_2 x_{i2} + \dots + \beta_p x_{ip}) \geq M(1 - \xi_i)   \qquad\forall i = 1..N$
		\item $\xi_i \geq 0, \sum_{i=1}^{N}\xi_i \leq \tilde{C}$
		\item where $\xi_i$ is a slack variable to allow instance $i$ to lie on the wrong side of the margin
		\item $\tilde{C}$ is the total budget for misclassifications
	\end{itemize}
\end{itemize}
Equivalent technical optimisation, so called "dual form"
\begin{itemize}[noitemsep,nosep]
	\item maximise\\
	$L_D = \sum_{i=1}^{N} \alpha_i - \frac{1}{2} \sum_{i=1}^{N}\sum_{k=1}^{N} \alpha_i \alpha_k y_i y_k x_i^T x_k$
	\begin{itemize}
		\item subject to $0 \leq \alpha_i \leq C$ and $\sum_{i=1}^{N}\alpha_i y_i = 0$
		\item $\beta = \sum_{i=1}^{N} \alpha_i y_i x_i$ is minimised
	\end{itemize}
\end{itemize}
\end{minipage}
\begin{minipage}[t]{0.4\linewidth}
		\centering
		\includegraphics[width=0.8\linewidth]{img/support_vector_classifier_penalty}
\end{minipage}

\subsection{Support Vector Machines}
A possible solution for linearly completely non-separable data is to map the data into a higher-dimensional space. In this space a separating hyperplane may be defined. 

\subsubsection{Variable Transformation}
The same idea applies, transform a non-linearly separable input function in $\R^n$ to $\R^{n+k}, k\geq 1$ where it may be linearly separable.

\begin{center}
	\includegraphics[keepaspectratio,width=0.6\linewidth]{variable_transformation.png}
\end{center}

The power of feature space transformation lies in the fact, that with an appropriately chosen feature space of sufficient dimensionality any consistent training set can be made separable.
\begin{equation*}
	\left\langle \begin{pmatrix}x_i\\x_i^2\end{pmatrix} \cdot\begin{pmatrix}x_{i'}\\x_{i'}^2\end{pmatrix} \right\rangle = x_i x_{i' } + x_i^2 x_{i' }^2 := K(x_i,x_{i'})
\end{equation*}
$K$ is known as a \textbf{kernel} function. It can be used to efficiently compute feature space transformations. Instead of calculating the inner product, the kernel $K(\cdot,\cdot)$ gets calculated. As soon as some $K$ is used, the resulting SVC becomes a SVM.

The following Kernels are common
\begin{itemize}[noitemsep]
	\item {\color{DodgerBlue2} Identity (inner product)}
	\begin{equation*}
		K(x_i,x_{i'}) = \sum_{j=1}^{p} x_{ij}x_{i'j}
	\end{equation*}
	this yields the standard support vector classifier
	\item {\color{DodgerBlue2} Polynomial of degree $d$}
	\begin{equation*}
		K(x_i,x_{i'}) = \left(1 + \sum_{j=1}^{p} x_{ij}x_{i'j}\right)^d
	\end{equation*}
	the polynomial kernel of degree $d = 1$ is just the identity kernel
	\item {\color{DodgerBlue2} Radial basis or Gaussian}
	\begin{equation*}
		K(x_i,x_{i'}) = e^{\left( -\gamma \sum_{j=1}^{p}(x_{ij} - x_{i'j})^2 \right)}
	\end{equation*}
	the Gaussian kernel computes a very local neighbourhood, with $\gamma$ being the hyper-parameter controlling the width of that neighbourhood. A low $\gamma$ yields a great width. The feature space spanned by the Gaussian kernel is implicit and infinite-dimensional.
\end{itemize}

\subsection{Multiclass Classification With Support Vector Machine}
With one versus rest or all classification, the binary classifier SVM can be adapted to choose between multiple classes.
\begin{center}
	\includegraphics[width=0.8\linewidth]{img/support_vector_machine_multiclass}
\end{center}
Assume $k$ classes, learn all $\frac{k(k-1)}{2}$ pairwise comparisons and classify unknown instances by majority vote among all pairwise comparisons.

\section{Ensemble Methods}
The goal of ensembles is to increase performance by \textbf{combining multiple complementary classifiers}. The intuition behind it is to build different experts and let them decide by vote. This is {\color{Green3} very effective in practice} and has {\color{Green3}good theoretical guarantees}, but while being {\color{Green3} easy to implement} the result {\color{Firebrick3} lacks in transparency}, {\color{Firebrick3}interpretability} and a {\color{Firebrick3} compact representation}.

The formal problem description is: Given $T$ binary classification hypotheses $(h_1,\dots, h_T)$, find a combined classifier with better performance of the form
\begin{equation*}
	\hat{h}(x) = \text{sign}\left( \sum_{t=1}^{T} \alpha_t h_t (x) \right)
\end{equation*}
For regression the average is used.

Ensembles work because they average the error, in statistical terms the bias remains equal but the variance is reduced.

\subsection{AdaBoost}
\subsubsection{Boosting}
The general idea is to boost the performance of weak learners iteratively by making misclassified examples more important and then combining hypotheses. This way, each stage additively corrects shortcomings of previous stages by reweighing and majority voting on the result. The \textbf{Ada}ptive \textbf{Boost}ing algorithm works by
\begin{minted}[escapeinside=||,mathescape=true,fontsize=\small]{python}
initialize weights: |$w_i := \frac{1}{N}$|
for |$ t:=1..T $|
	|$ h_t $| := train decision stump on the |$ x_i $|, weighted by the |$ w_i $|
	|$ \varepsilon_t $| := |$ \frac{\sum_{i=1}^{N} w_i \cdot I(y_i\neq h_t(x_i) }{\sum_{i=1}^{N} w_i} $| # compute error, $ I() $ is the identity function
	|$ \alpha_t $| := |$ \log\left(\frac{1-\varepsilon_t}{\varepsilon_t} \right) $| # compute influence of weak learner
	|$ w_i $| := |$ w_i\cdot e^{\alpha_t\cdot I(y_i \neq h_t(x))} $| # increase weight by exp(Influence) in case of error
return |$ \hat{h} $| := |$ \text{sign}\left(\sum_{t=1}^T \alpha_t\cdot h_t(x)\right) $| # majority vote
\end{minted}

\section{Bias, Variance and Learning Curves}

\begin{center}
	\includegraphics[width=0.8\linewidth]{img/bias_variance_tradeoff}
\end{center}

\subsection{Polynomial Regression using sklearn.pipeline}
\begin{minted}[linenos]{python}
from sklearn.pipeline import make_pipeline
from sklearn.linear_model import LinearRegression
from sklearn.preprocessing import PolynomialFeatures

model = make_pipeline(PolynomialFeatures(degree),
LinearRegression())
model.fit(x[:, np.newaxis], y)
ax.plot(x_test, model.predict(x_test[:, np.newaxis]), '-b')
\end{minted}

\subsection{Bias-Variance-Decomposition}
When the relationship between the predictor variable $X$ and the response variable $Y$ is written as
\begin{equation*}
	Y=f(X) + \epsilon \qquad \epsilon\sim\N{0,\sigma_{\epsilon}}
\end{equation*}
Then the expected value of the quadratic error can be written as
\begin{equation*}
	\text{SE}(x) = \ev{(Y - \hat{f}(x))^2}
\end{equation*}
Transformed
\begin{equation*}
	\text{SE}(x) = \underbrace{\left(\ev{\hat{f}(x)} - f(x)\right)^2}_{\text{Bias}^2} + \underbrace{\ev{\left(\hat{f}(x) - \ev{\hat{f}(x)}\right)^2}}_{\text{Variance}} + \underbrace{\sigma_{\epsilon}^2}_{\text{irreducible error}}
\end{equation*}

\subsection{The No-Free-Lunch Theorem (Probably Approximately Correct Theory)}
\begin{theorem}
	For every learner, there exists a task on which it fails, even though that task can be successfully learned by another learner.
\end{theorem}
Intuitively, as in the real world a learner is never trained on the whole data that exists, but only a subset of that data.

\subsection{Cross-Validation}
\begin{center}
	\includegraphics[width=0.7\linewidth]{img/cross_validation_hyperparameter_tuning}
\end{center}
\subsubsection{Stratified k-Fold Cross Validation for Unequal Classes}
In cases where classes are unevenly distributed, {\color{DodgerBlue3} stratified k-fold cross-validation} reduces bias and variance of the estimation. The implementation takes care to ensure that the classes are distributed approximately equally in the subsets. This ensures that the individual subsets are representative of the class distribution in the training data set.

\begin{minted}[linenos,fontsize=\small]{python}
from sklearn.model_selection import validation_curve

degrees = np.arange(1, 14)
model = make_pipeline(PolynomialFeatures(), LinearRegression())
# The parameter to vary is the "degrees" on the pipeline step "polynomialfeatures"

train_scores, validation_scores = validation_curve(
	model, x[:, np.newaxis],
	y,
	param_name='polynomialfeatures__degree',
	param_range=degrees)
\end{minted}

\subsubsection{Cross-Validated Grid Search in Python}
A grid search consists of
\begin{itemize}[nosep,noitemsep]
	\item an \textbf{estimator}
	\item a \textbf{parameter space}
	\item a method for searching or \textbf{sampling} candidates;
	\item a \textbf{cross-validation scheme}
	\item a \textbf{score function}
\end{itemize}
\begin{minted}[linenos,fontsize=\small]{python}
from sklearn import svm, datasets
from sklearn.model_selection import GridSearchCV

iris = datasets.load_iris()

parameters = {'kernel':('linear', 'rbf'), 'C':[1, 10]}
svc = svm.SVC(gamma="scale")

clf = GridSearchCV(svc, parameters, cv=5)
clf.fit(iris.data, iris.target)

sorted(clf.cv_results_.keys())
\end{minted}

\subsection{Learning Curves}
\begin{center}
	\includegraphics[width=0.9\linewidth]{img/learning_curve_bias_variance}
\end{center}
A \textbf{high bias} points to a systemic error or underfit, choose another model to approximate the real function. A \textbf{high variance} points to instability or overfit, getting more data or use more regularisation.
\subsection{Regularisation}
Ridge-Regression ($L_2$-Regression) \textbf{distributes weight} across related features and has an analytic solution
\begin{equation*}
	R_{\text{Ridge}}(\theta) = \sum_{i=1}^{d} \theta_i^2
\end{equation*}
LASSO Regression ($L_1$-Regression) encourages sparsity by setting some weights to zero and is used to select \textbf{informative features}
\begin{equation*}
	R_{\text{LASSO}}(\theta) = \sum_{i=1}^d\abs{\theta_i}
\end{equation*}

\subsection{Reduction of the Model Complexity}
Increasing the complexity of a model usually increases its variance and lowers its bias. Conversely, a lower complexity of the model increases its bias and lowers its variance. Therefore, this is called equilibrium. Regularising the model, by restricting it in some manner, is one way to avoid overfitting: The less freedom the model has, the more difficult it becomes to overfit the data. For example, a polynomial model can easily be regularised by reducing the degree of the polynomial.

\subsubsection{Minimisation of the Loss Function with Regularisation}
\begin{align*}
	L(\theta) &= \sum_{i=1}^{N} \left( y_i - \hat{y}_i(\theta) \right)^2 + \lambda \norm{\theta}_R\\
	\norm{\theta}_R:\qquad \norm{\theta}_2^2 &= \sum_{i=1}^{n} (\theta_k)^2\qquad \text{Ridge-Regression}\\
	\norm{\theta}_1 &= \sum_{i=1}^{n} \abs{\theta_k}\qquad \text{LASSO-Regression}\\
\end{align*}
with $\lambda$ being the Regularisation hyperparameter. By assigning larger values to the hyper parameter $\lambda$, we increase the regularisation strength and reduce the weights of our model. LASSO is an approach that can lead to sparse models. Depending on the regularisation strength, certain weights can become zero, which means that LASSO can also function as a \textbf{monitored feature selection procedure}.

\section{System Design}
Machine Learning systems as a whole are pipelines composed of individual components, that can be developed collaboratively as a team.

\subsection{Ceiling Analysis}
\begin{enumerate}[noitemsep,nosep]
	\item {\color{Tomato2} Baseline}: Measure the performance of the complete pipeline
	\item Replace {\color{Goldenrod1} first component} with the ground truth (perfect result for this stage) and measure performance of the system
	\item Replace {\color{OliveDrab3} next component} with the ground truth and measure performance of the system
	\item \qquad $\vdots$
\end{enumerate}
\begin{figure}[H]
	\centering
	\includegraphics[width=\linewidth]{img/ceiling_analysis.png}
	\caption{Based on this analysis to improve Text Detection and Character Recognition makes sense, Character Segmentation only improves the overall performance by 1\% and is thus not worth the investment}
\end{figure}

\section{Feature Engineering}
A feature is an individual measurable property of a phenomenon being observed.

\subsubsection{Exploratory Data Analysis}
Whether collected by your workgroup or obtained from someone else, raw data is seldom ready for immediate analysis. Exploratory data analysis can help in discovering important anomalies, identifying limitations in the collection process, and to better inform subsequent goal oriented analysis.

Before generating features for the model, identify key properties of the data including structure, granularity, faithfulness, temporality and scope. This can be done by preparing, analysing and visualising the data using Histograms, QQ-Plots, Scatter-matrices and Heatmaps.

\subsubsection{Data Cleansing}
\begin{minted}{python}
# drop duplicates, drop not available
df.drop_duplicates()
df.dropna()

# replace the cells without any data in the column with the value 'None'
df = df.astype(object).where(pd.notnull(df),None)
weekday_counts = df.groupby('weekday').aggregate(sum)

# replace * in Name with empty string:
df['Name']=df['Name'].apply(lambda x:str(x).replace('*',''))

# convert dates to datetime, keeping only the month
df['MonthYear'] = pd.to_datetime(df['MonthYear'])
df['MonthYear'] = df['MonthYear'].apply(lambda x: x.date())
\end{minted}

\subsection{Feature Preparation}
\begin{itemize}
	\item \textbf{Data Cleaning}\\
	Homogenise missing values and different types of in the same feature, fix input errors, types, etc.
	\item \textbf{Aggregating / Pivoting}\\
	necessary when the entity to model is an aggregation from the provided data.
	\item \textbf{Imputation of missing values}\\
	Strategies: mean, median, mode, using a model
	\item \textbf{Binarisation}\\
	transform discrete or continuous numeric features into binary features
	\item  \textbf{Binning}\\
	Fixed width, adaptive quantile binning
	\item \textbf{Transformation}\\
	Compress the range of large numbers or expand the range of small numbers.
	\item \textbf{Scaling and normalisation}
	\item \textbf{Generate Interaction features}
\end{itemize}

\subsubsection{Box-Cox-Transformation}
\begin{minipage}{0.5\linewidth}
	The Box-Cox Transformation can be used to symmetrize positive numerical features
	\begin{equation*}
		\tilde{x}_i = \left\{ \begin{matrix}
		\frac{x_i^\lambda - 1}{\lambda} & \text{if }\lambda\neq 0\\
		\log(x_i) & \text{otherwise}
		\end{matrix} \right.
	\end{equation*}
\end{minipage}
\begin{minipage}{0.5\linewidth}
	\begin{center}
		\includegraphics[width=0.9\linewidth]{img/boxcox_lambdas}
	\end{center}
\end{minipage}
\begin{minted}{python}
from sklearn.preprocessing import PowerTransformer

# select numeric features
weather_num = weather.select_dtypes(include =["float"]) #
# 'yeo-johnson' (default), works with positive and negative values
# 'box-cox', only works with strictly positive values
PowTrans = PowerTransformer(method='yeo-johnson')
# fit and transform data
weather_num_t = PowTrans.fit_transform(weather_num)
\end{minted}

\subsection{Feature Preparation}
\subsubsection{One-Hot Encoding}
\begin{minted}{python}
from sklearn.preprocessing import OneHotEncoder
enc = OneHotEncoder(handle_unknown='ignore')
X = [['Male', 1], ['Female', 3], ['Female', 2]]
enc.fit(X)
\end{minted}

\subsubsection{Bin Counting}
\begin{minted}{python}
rnd = np.random.RandomState(42); X = rnd.uniform(-3, 3, size=100)
y = np.sin(X) + rnd.normal(size=len(X)) / 3
X = X.reshape(-1, 1)
enc = KBinsDiscretizer(n_bins=10, encode='onehot')
X_binned = enc.fit_transform(X)
\end{minted}

\subsubsection{Label Encoding}
\begin{minted}{python}
le = preprocessing.LabelEncoder()
le.fit(["paris", "paris", "tokyo", "amsterdam"])
list(le.classes_)
> ['amsterdam', 'paris', 'tokyo']
le.transform(["tokyo", "tokyo", "paris"])
> array([2, 2, 1])
\end{minted}

\subsubsection{Note on NLP}
This section is covered in 'Analysis of Text Data' in much more detail, and has thus been omitted.

\subsection{Feature Selection}
The reason to use feature selection is reducing the number of features, to reduce overfitting and improve the generalization of models, and to gain  a better understanding of the features and their relationship to the response variables.

\subsection{Feature Selection using Linear Models and Regularisation}
Utilise machine learning models for feature ranking as many have some inherent internal ranking feature. Even simple linear regression
models can work when the data is not very noisy and the features are independent.

To avoid a multicollinearity problem either L1 (Lasso) or L2 (Ridge) regularisation should be used.

\begin{minted}{python}
from sklearn.linear_model import Lasso
from sklearn.preprocessing import StandardScaler
from sklearn.datasets import load_boston

boston = load_boston()
scaler = StandardScaler()
X = scaler.fit_transform(boston['data'])
Y = boston['target']
names = boston['feature_names']

lasso = Lasso(alpha=.3)
lasso.fit(X,Y)
\end{minted}
\texttt{Lasso model: -3.707 * LSTAT + 2.992 * RM + -1.757 * PTRATIO + -1.081 * DIS\\
	+ -0.7 * NOX + 0.631 * B + 0.54 * CHAS + -0.236 * CRIM + 0.081 * ZN\\
	+ {\color{red} -0.0 * INDUS} + {\color{red} -0.0 * AGE} + {\color{red}0.0 * RAD} + {\color{red}-0.0 * TAX}}
L2 or Ridge regularisation squares the coefficients in the penalty expression. This forces the coefficient values to be spread out more equally.

\subsection{Tree-based Methods for Feature Selection}
Random forest are popular methods due to their good accuracy, robustness and ease of use. Additionally, they provide a mean decrease purity and mean decrease accuracy.

\subsubsection{Information Gain}
\begin{equation*}
	\text{Entropy }H(T) = -\sum_{i=1}^{K} p_i\cdot\log_2(p_i)
\end{equation*}
Where $p_i$ represent the percentage of each class present in the child node that results from a split in the tree.
\begin{equation*}
	\text{Information Gain } \text{IG}(T,a) = H(T) - H(T|a)
\end{equation*}
\begin{equation*}
	\text{IG}(T,a) = -\sum_{i=1}^{K} p_i\cdot\log_2(p_i) - \sum_a p(a)\cdot\sum_{i=1}^{K} p(i|a)\cdot\log_2[p(i|a)]
\end{equation*}

\subsection{Recursive Feature Elimination (RFE)}
Given an external estimator that weighs features, RFE selects features by recursively considering smaller sets of features.

\section{Bayes' Theorem}
\subsection{Multivariate Gaussian Distribution}
Gaussian defined over a vector x of continuous variables in a D-dimensional space with mean vector $\mu$ and covariance matrix $\bm{\Sigma}$, where $ \abs{\bm{\Sigma}} $ is the determinant of $\bm{\Sigma}$.
\begin{equation*}
	p(\bm{x}) = \N{\bm{x}|\mu, \bm{\Sigma}} = \frac{1}{(2\pi)^{\frac{D}{2}} \norm{\bm{\Sigma}}^{\frac{1}{2}}} \exp\left\{ -\frac{1}{2} (\bm{x} - \mu)^T \bm{\Sigma}^{-1} (\bm{x} - \mu) \right\}
\end{equation*}
The quadratic form in the argument of the exponential is called \emph{Mahalanobis distance}
\begin{equation*}
	\bm{\Delta} = (\bm{x} - \mu)^T \bm{\Sigma}^{-1} (\bm{x} - \mu)
\end{equation*}
Without loss of generality, we can assume that $\bm{\Sigma}$ is symmetric with real eigenvalues and an orthonormal set of eigenvectors $\mu_i$.

\subsection{Probability Distribution}
Since $\bm{\Sigma}$ is real and symmetric, the eigenvectors $\lambda_i$ are real and the eigenvectors can be chosen from an orthonormal set, so that
\begin{align*}
	\textbf{u}_i^T \textbf{u}_j &= \delta_{ij}\\
	\bm{\Sigma} &= \sum_{i=1}^{D} \lambda_i \textbf{u}_i\textbf{u}_i^T\\
	\bm{\Sigma}^{-1} &= \sum_{i=1}^{D} \frac{1}{\lambda_i} \textbf{u}_i\textbf{u}_i^T\\
	\intertext{The quadratic form becomes}
	\Delta^2 &= \sum_{i=1}^{D}\frac{y_i^2}{\lambda_i}\\
	y_i &= \textbf{u}_j^T (\textbf{x} - \mu)
\end{align*}
This describes a D-dimensional ellipsoid with main axes $\sqrt{\lambda_i}$. The determinant is given by the product of the eigenvalues
\begin{equation*}
	\abs{\bm{\Sigma}}^{\tfrac{1}{2}} = \prod_{j=1}^{D} \lambda_j^{\tfrac{1}{2}}
\end{equation*}

\begin{definition}
	\begin{align*}
		\intertext{Expected Value}
		\ev{\textbf{x}} &= \bm{\mu}\\
		\intertext{Correlation}
		\ev{\textbf{xx}^T} &= \bm{\mu\mu}^T + \bm{\Sigma}\\
		\intertext{Covariance}
		\text{Cov}[\textbf{x}] &= \ev{(\textbf{x}-\ev{\textbf{x}})(\textbf{x}-\ev{\textbf{x}})^T} = \bm{\Sigma}
	\end{align*}
\end{definition}

Partition $\textbf{x}$ into two disjoint subsets $\textbf{x}_a$ and $\textbf{x}_b$. Without loss of generality, $\textbf{x}_a$ can be taken to form the first $M$ components of $\textbf{x}$ with $\textbf{x}_b$ comprising of the remaining $D-M$ components
\begin{alignat*}{4}
	\textbf{x}&=\begin{pmatrix}\textbf{x}_a\\\textbf{x}_b\end{pmatrix} &\qquad& \bm{\mu} &= \begin{pmatrix}\bm{\mu}_a\\\bm{\mu}_b\end{pmatrix} &\qquad& \bm{\Sigma} &= \begin{pmatrix}\bm{\Sigma}_a\\\bm{\Sigma}_b\end{pmatrix}
\end{alignat*}
In most situations it is more convenient to work with the \textbf{inverse of the covariance matrix}, which is called the \textbf{precision matrix}
\begin{align*}
	\bm{\Lambda} &\equiv \bm{\Sigma}^{-1}\\
	\bm{\Lambda} &= \begin{pmatrix}
		\bm{\Lambda}_{aa} & \bm{\Lambda}_{ab}\\
		\bm{\Lambda}_{ba} & \bm{\Lambda}_bb
	\end{pmatrix}
\end{align*}

% TODO Conditional probability distribution & Marginal probability distribution

% TODO Conditional Gaussian Distributions

\subsection{Belief Networks or Bayesian Networks}
A belief network is a directed acyclic graph (DAG) in which each node has associated the conditional probability of the node given its parents. The joint distribution is obtained by taking the product of the conditional probabilities
\begin{center}
	\includegraphics[width=0.7\linewidth]{img/belief_network}
\end{center}
$ \bm{X} \incond \bm{Y} | \bm{Z}$ denotes that two sets of variables $\bm{X}$ and $\bm{Y}$ are independent of each other given the state of the set of variables $\bm{Z}$. This means that
\begin{align*}
	p(\bm{X},\bm{Y}|\bm{Z}) &= p(\bm{X}|\bm{Z})\cdotp(,\bm{Y}|\bm{Z})\\
	p(\bm{X}|\bm{Y},\bm{Z}) &= p(\bm{X}|\bm{Z})
\end{align*}
for all states of $\bm{X},\bm{Y},\bm{Z}$. In case the conditioning set is empty we may also write
\begin{equation*}
	\bm{X} \incond \bm{Y}
\end{equation*}
in which case $\bm{X}$ is unconditionally independent of $\bm{Y}$.
\begin{center}
	\includegraphics[width=0.7\linewidth]{img/conditional_independence}
\end{center}
In (d) the variables $A,B$ are conditionally dependent given $C$
\begin{equation*}
	P(A,B|C) \propto p(C|A,B)p(A)p(B)
\end{equation*}

\subsubsection{Collider}
A collider contains two or more incoming arrows along a chosen path.
\begin{center}
	\includegraphics[width=0.8\linewidth]{img/conditional_independence_collider}
\end{center}

\subsection{Bayes' Theorem}
\begin{center}
	\includegraphics[width=0.6\linewidth]{img/bayes_theorem}
\end{center}

\section{Gaussian Processes}
Is referred to as the \textbf{infinite-dimensional extension of the multivariate normal distribution}. When we are working with Gaussian Processes the intuition is that we observe some finite-dimensional subset of infinite-dimensional data, and this finite subset follows a multivariate normal distribution, as would every finite subset.

Gaussian random variables are useful in machine learning and statistics for two main reasons.
\begin{enumerate}
	\item They are extremely common when modelling 'noise' in statistical algorithms
	\item Gaussian random variables are convenient for many analytical manipulations, because many of the integrals involving Gaussian distributions that arise in practice have \textbf{simple closed form solutions}
\end{enumerate}

\subsection{Bayesian Regression}
\begin{align*}
	\intertext{Parameter Posterior}
	p(\theta|\mathcal{D}) &= \frac{p(\mathcal{D}|\theta)\cdot p(\theta)}{\int_{\theta'} p(\mathcal{D}|\theta)\cdot p(\theta) \diff \theta'}
	\intertext{Posterior Predictive Distribution}
	p(y_{*}|x_{*}\mathcal{D}) &= \int_{\theta} p(x_{*}|y_{*},\theta)\cdot p(\theta|\mathcal{D})\diff\theta
\end{align*}
\texttt{pymc3} is a comfortable package for Bayesian Inference in python.
\begin{minted}{python}
from pymc3 import *

with Model() as model:
	# model specifications in PyMC3 are wrapped in a with-statement
	# Define priors (#tau = precision = 1/variance = 1/sigma^2)
	sigma = HalfCauchy('sigma', beta=10, testval=1.)
	intercept = Normal('Intercept', 0, tau=1/20**2)
	x_coeff = Normal('x', 0, tau=1/20**2)
	
	# Define likelihood
	likelihood = Normal('y', mu=intercept + x_coeff * x, tau=1/sigma**2, observed=y)
	
	# Inference!
	trace = sample(3000, cores=2)
	
	# draw 3000 posterior samples using NUTS sampling
	plt.figure(figsize=(7, 7))
	traceplot(trace[100:]); plt.tight_layout();
\end{minted}

\subsection{Gaussian Processes}
\begin{equation*}
	h(\bm{x})\sim \N{\bm{m}(\bm{x}), \bm{k}(\bm{x},\bm{x}')}
\end{equation*}
\textbf{Gaussian Processes} are the extension of multivariate Gaussians to infinite-sized collections of real-valued variables. Gaussian processes can be thought of as distributions not just over random vectors but in fact distributions over random functions. A Gaussian process is a stochastic process such that any finite sub-collection of random variables has a multivariate Gaussian distribution.

In particular, a collection of random variables $ \{h(x): x\in\mathcal{X}\} $ is said to be drawn from a Gaussian process with \textbf{mean function $\bm{m(\cdot)}$} and \textbf{covariance function} $\bm{k(\cdot,\cdot)}$ if for any finite set of elements $x_1,x_2,\dots,x_m\in\mathcal{X}$ and the associated finite set of random variables $h(x_1), h(x_2),\dots,h(x_m)$ have a distribution

\begin{center}
	\includegraphics[width=0.8\linewidth]{img/gaussian_processes}
\end{center}

\subsection{Non-Parametric Gaussian Process Models}
% TODO NP GPM


\section{Dimensionality Reduction}
The "Curse of Dimensionality" is the problem that inputs are high dimensional when working with data from speech, images, genomes and other sources. One of the characteristics of high-dimensional data is that the number of dimensions is comparable or larger than the number of samples. This has the consequence that the sample complexity of function approximation can grow exponentially.

\begin{figure}[H]
	\centering
	\includegraphics[width=0.6\linewidth]{img/curse_dimensionality}
	\caption{The curse of dimensionality is well illustrated by a subcubical neighborhood for uniform data in a unit cube. The figure on the right shows the side-length of the subcube needed to capture a fraction r of the volume of the data, for different dimensions p. In ten dimensions we need to cover 80\% of the range of each coordinate to capture 10\% of the data.}
	\label{fig:cursedimensionality}
\end{figure}

High dimensional data is very sparse: most training instances are likely to be far away from each other making predictions much harder, greater risk of overfitting, joint probabilities are hard to fit.

\subsection{Dimensionality Reduction}
Dimensionality reduction is the process of reducing the number of random variables under consideration by obtaining a set of principal variables.
\begin{itemize}
	\item \textbf{Feature Reduction} approaches try to find a subset of the original variables (also called features or attributes).
	\begin{itemize}
		\item filter strategy, \emph{information gain}
		\item wrapper strategy, \emph{search guided by accuracy}
		\item embedded strategy, \emph{features are selected to add or be removed while building the model based on the prediction errors}
	\end{itemize}
	\item \textbf{Feature Extraction} transforms the data in the high-dimensional space to a space of fewer dimensions
\end{itemize}

\newpage
\noindent
\begin{tabularx}{\linewidth}{m{0.2\linewidth} m{0.6\linewidth}}
	\multicolumn{2}{l}{\textbf{Linear Techniques}}\\
	\hline
	&\\[-0.5em]
	\includegraphics[width=\linewidth]{img/input_data_variance} & Several methods (starting from PCA) attempt to identify the low- dimensional subspace that captures the largest fraction of the input data variance\\
	\includegraphics[width=\linewidth]{img/kernel_method} & This is equivalent to finding the best L2-norm approximation of the Gram matrix, and generalises to \textbf{kernel methods}\\
	& \\
	\multicolumn{2}{l}{\textbf{Manifold Techniques}}\\
	\hline
	&\\[-0.5em]
	\includegraphics[width=\linewidth]{img/multi-dimensional_scaling} & \textbf{Multi-dimensional scaling} and related approaches attempt to reproduce the similarity between high-dimensional data points in low dimension\\
	\includegraphics[width=\linewidth]{img/embedding_methods} & \textbf{Embedding methods} explicitly try to preserve local relations between points, under the assumption that they lie on a (locally) low-dimensional manifold
\end{tabularx}

\subsection{Principal Component Analysis}
The main linear technique for dimensionality reduction, principal component analysis PCA, performs a linear mapping of the data to a lower-dimensional space in such a way that the variance of the data in the low-dimensional representation is maximized.

The eigenvectors that correspond to the largest eigenvalues (the principal components) can now be used to reconstruct a large fraction of the variance of the original data. PCA identifies the d-dimensional hyperplane that lies closest to the data and then projects the data on it (d is a hyperparameter).

\begin{center}
	\includegraphics[width=0.7\linewidth]{img/input_covariance_matrix}
\end{center}

\begin{align*}
	\textbf{C}_ij &= \left\langle (\textbf{X}_i - \left\langle \textbf{X}_i \right\rangle) (\textbf{X}_j - \left\langle \textbf{X}_j \right\rangle) \right\rangle\\
	\left\langle \textbf{X}_i \right\rangle &= 0 \Rightarrow \textbf{C} = \textbf{X}^T \textbf{X}
\end{align*}
It turns out that the optimization corresponds to a diagonalization of the covariance matrix $  \textbf{C} = \textbf{U}\textbf{D}\textbf{U}^T$. PCA is a rotation or reflection so that the first principal component has the highest variance, the second is orthogonal to the first and has the second highest variance and so forth. $\textbf{X}$ is the $n\times p$-dimensional data matrix column centred.
\begin{center}
	\includegraphics[width=0.7\linewidth]{img/PCA_X}
\end{center}

\subsubsection{Explained Variance Ratio $P_k$}
\begin{minipage}{0.55\linewidth}
	The percentage the first $k$ dimensions contribute to the variance.
	\begin{equation*}
	P_k = \frac{\sum_{i=1}^k \text{var}\left(\textbf{X}_i\right)}{\sum_{j=1}^p \text{var}\left(\textbf{X}_j\right)} = \frac{\sum_{i=1}^{k} \lambda_i}{\sum_{j=1}^{p} \lambda_j}
	\end{equation*}
\end{minipage}
\begin{minipage}{0.4\linewidth}
	\begin{center}
		\includegraphics[width=\linewidth]{img/explained_variance_ratio}
	\end{center}
\end{minipage}


% TODO PCA

% TODO PCA Rotation

\section{Clustering}
Clustering is finding groups in data. Given $n$ data points, separate them into $K$ clusters ($n$: number of data points, $K$: number of clusters, $\Delta$: a partition $\Delta = \{C_1, C_2,\dots, C_K\}$, $\mathcal{L}(\Delta)$: loss of $\Delta$ to be minimised). There is a differentiation between hard clustering, where each data point is assigned to a unique cluster, and soft clustering, where each data point $i$ is assigned a probability that it is in cluster $k:=\{\gamma_{ki}\}_{k=1:K}$. Clustering is related to vector quantisation.

\subsection{Taxonomy of Clustering}
\begin{itemize}
	\item \textbf{Parametric clustering}: $K$ known
	\item \textbf{Non-parametric clustering}: $K$ determined by algorithm
\end{itemize}
\textbf{Hierarchical Clustering} (HCA) seeks to build a hierarchy of clusters. Strategies for hierarchical clustering generally fall into two types
\begin{itemize}
	\item \textbf{Agglomerative}, a bottom-up approach where each observation starts in its own cluster, and pairs of clusters are merged the longer the algorithm is run
	\item \textbf{Divisive}, a top-down approach where all observations start in one cluster and splits are performed recursively as the algorithm moves down the hierarchy
\end{itemize}
In hierarchical clustering, clusters have a tree like structure or a \emph{parent child relationship}. Here, the two most similar clusters are combined together and continue to combine until all objects are in the same cluster.

\subsubsection{Other Distinctions}
\begin{itemize}
	\item \textbf{Exclusive versus non-exclusive}, where in non-exclusive clusterings points may belong to multiple clusters
	\item \textbf{Fuzzy versus non-fuzzy}, where in fuzzy clustering a point belongs to every cluster with some weight between 0 and 1, who must sum up to 1. Probabilistic clustering has similar characteristics
	\item \textbf{Partial versus complete}, where in the partial case not the whole dataset is used
	\item \textbf{Heterogeneous versus non-heterogeneous}, where clusters can have widely different sizes, shapes or densities
\end{itemize}

\subsection{Hierarchical Clustering}
In order to decide which clusters should be combined or where a cluster should be split, a measure of dissimilarity between sets of observations is required. In most methods of hierarchical clustering, this is achieved by use of an appropriate \textbf{metric} (a measure of distance between pairs of observations), and a \textbf{linkage criterion} which specifies the dissimilarity of sets as a function of the pairwise distances of observations in the sets.

In agglomerative clustering there are $N$ data points at the start that initially form $N$ clusters. The two clusters with the smallest linkage are fused together to form $N-1$ clusters. This is repeated until there is only one single cluster.
\begin{equation*}
	(i,j)_{\text{fused}} = \underset{i\neq j}{\text{argmin}} \left\{ D_{\text{link}})(C_i,C_j) \right\}
\end{equation*}

\vspace{1em}
\noindent
A \textbf{metric} d(x,y) is a generalised distance measure that follows these axioms
\begin{tabularx}{\linewidth}{l X}
	1. Non-negativity & $d(x,y)\geq 0$\\
	2. Coincidence & $d(x,y) = 0 \Leftrightarrow x=y$\\
	3. Symmetry & $d(x,y) = d(y,x)$\\
	4. Triangle inequality & $d(x,y) + d(y,z) \geq d(x,z)$
\end{tabularx}
\begin{itemize}
	\item $L_p$ metric: \quad $d_p(x,y) = \norm{x-y}_p = \sqrt[p]{\sum_i \abs{x_i - y_i}^p}$
	\item $L_\infty$ metric: \quad $\norm{x}_\infty = \max_i\left\{\abs{x_i}\right\}$
	\item $L_1$ metric: \quad $d_1(x,y) = \norm{x-y}_1 = \sum_i\abs{x_i - y_i}$
\end{itemize}
\clearpage
The linkage criterion determines together with a metric $d(x,y)$ when two clusters $A$ and $B$ should be merged together in hierarchical clustering (fusion criterium).
\begin{tabularx}{\linewidth}{m{0.35\linewidth} m{0.35\linewidth} m{0.2\linewidth}}
	\textbf{Names} & \textbf{Formula} & \textbf{Plot}\\
	\hline
	&&\\[-0.5em]
	\textbf{Maximum} or complete-linkage clustering & $ \max\left\{ d(a,b):a\in A, b\in B \right\} $ & \includegraphics[width=\linewidth]{complete_linkage}\\
	\textbf{Minimum} or single-linkage clustering & $ \min\left\{ d(a,b):a\in A, b\in B \right\} $ & \includegraphics[width=\linewidth]{single_linkage}\\
	\textbf{Mean}, average-linkage clustering or UPGMA & $ \frac{1}{\abs{A}\cdot\abs{B}}\sum_{a\in A}\sum_{b\in B} d(a,b) $ & \includegraphics[width=\linewidth]{img/average_linkage}\\
	\textbf{Ward} linkage & $D_{\text{Ward}}(A,B) := \frac{d(\bar{a},\bar{b})^2}{\dfrac{1}{\abs{A}} + \dfrac{1}{\abs{B}}}$; increase of variance when fusing $A$ and $B$ & \\
	\textbf{Maximum} or complete-linkage clustering & $ \norm{c_s - c_t} $, where $c_s$ and $c_t$ are the centroids of clusters $s$ and $t$ & \includegraphics[width=\linewidth]{img/centroid_linkage}
\end{tabularx}

\subsubsection{Basic Form of an Agglomerative Linkage Algorithm}
\textbf{Input}:
\begin{itemize}[nosep]
	\item Distance matrix $D$ between data points
	\item function \texttt{dist} to compute a distance between clusters (which usually takes $D$ as an input)
\end{itemize}
\textbf{Initialisation}: Clustering $\mathcal{C}^{(0)} = \left\{ C_1^{(0)}, C_2^{(0)},\dots, C_n^{(0)} \right\} = \{i\}$
\begin{itemize}
	\item \textbf{While} the current number of clusters is $>1$:
	\begin{itemize}[nosep]
		\item find the two clusters which have the \textbf{smallest distance (linkage)} to each other
		\item \textbf{merge} them to one cluster
	\end{itemize}
	\item \textbf{Output}: Resulting \textbf{dendogram}, which is a tree that represents the hierarchical division of the data set $O$ into ever smaller subsets.
\end{itemize}

\begin{minted}[linenos]{python}
# Import dendrogram and ward clustering from SciPy
from scipy.cluster.hierarchy import dendrogram, ward
X, y = make_blobs(random_state=0, n_samples=12)
#Perform ward clustering on the data in array X. The function ward in SciPy
returns an array with the #distances bridged in agglomerative clustering
linkage_array = ward(X)
# draw a dendrogram
dendrogram(linkage_array)
\end{minted}
\begin{center}
	\includegraphics[width=0.7\linewidth]{img/dendrogram}
\end{center}

\subsubsection{Divisive Clustering - Single Linkage Algorithm Based on Minimum Spanning Tree}
\textbf{Input}: Data $D = \{ x_i \}_{i=1:N}$ and number of clusters $K$
\begin{enumerate}
	\item Construct the Minimum Spanning Tree (MST) of $D$
	\item Delete the largest $(K-1)$ edges
\end{enumerate}
\textbf{Cost}:
\begin{align}
	\mathcal{L}(\Delta) &= \underset{k,k'}{\min}d(C_k, C_{k'})\\
	d(A,B) = \underset{x\in A, y\in B}{\min}\norm{x-y}
\end{align}
This method is sensitive to outliers, but the cost can be evaluated in polynomial time $O(n^2)$.

\subsection{K-Means}
\begin{figure}[H]
	\begin{subfigure}{0.45\linewidth}
		\centering
		\includegraphics[width=3cm]{kmeans01}
		\caption{$k$ initial "means" are randomly generated within the data domain}
	\end{subfigure}
	\hspace{\fill}
	\begin{subfigure}{0.45\linewidth}
		\centering
		\includegraphics[width=3cm]{kmeans02}
		\caption{$k$ clusters are created by associating every observation with the nearest mean. The partitions here represent the Voronoi diagram generated by the means.}
	\end{subfigure}
	\begin{subfigure}{0.45\linewidth}
		\centering
		\includegraphics[width=3cm]{kmeans03}
		\caption{The centroid of each of the $k$ clusters becomes the new mean}
	\end{subfigure}
	\hspace{\fill}
	\begin{subfigure}{0.45\linewidth}
		\centering
		\includegraphics[width=3cm]{kmeans04}
		\caption{The preceding steps are repeated until convergence has been reached}
	\end{subfigure}
\end{figure}

\subsubsection{Algorithm}
\begin{itemize}
	\item \textbf{Input}: Data $D = \{ x_i \}_{i=1:N}$ and number of clusters $K$
	\item \textbf{Initialise}: centers $\mu_1,\mu_2,\dots,\mu_k \in \R^d$ at random
	\item Iterate until convergence
	\begin{enumerate}
		\item \texttt{for} $i=1:n$\\
		$k(i) = \text{argmin}_k\norm{x_i - \mu_k}$ (assign points to cluster to get a new clustering)
		\item \texttt{for} $i=1:n$\\
		\begin{equation*}
			\mu_k = \frac{1}{\abs{C_k}}\sum_{i\in C_k} x_i
		\end{equation*}
		(recalculate centres)
	\end{enumerate}
	\item \textbf{Convergence}: if $\Delta$ does not change after iteration $m$, it will never change after that
\end{itemize}

\subsubsection{Cost Function}
The least-squares cost function is also called \textbf{distortion} (within cluster inertia $W$)
\begin{align*}
	\mathcal{L}(\Delta) &= \sum_{i=1}^{n}\norm{x_i - \mu_{k(i)}}^2\\
	&= \sum_{k=1}^{K}\sum_{i\in C_k} \norm{x_i - \mu_{k(i)}}^2
\end{align*}
The distortion can also be expressed as \textbf{sum of (squared) intra-cluster distances}
\begin{equation*}
	\mathcal{L}(\Delta) = \frac{1}{2} \sum_{k=1}^{K}\sum_{i\in C_k} \norm{x_i - x_j}^2 + \text{constant}
\end{equation*}

\subsubsection{Initialisation}
\textbf{Random Initialisation}
\begin{itemize}[nosep]
	\item Most common: randomly choose some data points as starting centers.
	\item Draw starting points randomly from $\R^d$
	\item Initialize the centers using the solution of an even simpler clustering algorithm
	\item Ideally have prior knowledge, for example that certain points are in different clusters
\end{itemize}

\subsubsection{More Variants}
\begin{itemize}
	\item \textbf{K-Medians}: The median is computed in each single dimension in the Manhattan-distance formulation of the k-medians algorithm
	\item \textbf{Weighted K-Means}: introduce weights for the individual data points
	\item \textbf{Kernel-K-Means}: the kernelized version of K-means
	\item \textbf{Soft K-Means}: no hard assignments, but soft assignments
\end{itemize}

\subsubsection{K-Medoid and K-Maxoid Clustering}
% TODO KMeans Methods

\section{Gaussian Mixture Models And The Expectation Maximisation Algorithm}
\subsection{The EM Algorithm: A General-Purpose, Unsupervised Learning Algorithm}
Expectation Maximisation is an iterative method to learn in the presence of unobserved variables. A typical hidden variable is some sort of group or cluster membership. This algorithm has good convergence guarantees.

\begin{enumerate}
	\item Start with a random initial hypothesis
	\item \textbf{E-Step}\\
	Estimate expected values of unobserved variables, assuming the current hypothesis holds
	\item \textbf{M-Step}\\
	Calculate new Maximum Likelihood (ML) estimate of hypothesis, assuming the expected values from (2) hold
	\item Iterate until convergence while always replacing old estimates with new ones
\end{enumerate}

\subsubsection{K-Means Revisited}
Assume that K is known. The optimisation that leads to the clusters is the following
\begin{flalign*}
	\underset{\textbf{z},\bm{\mu}}{\min}\mathcal{L}(\textbf{z},\bm{\mu}) = \sum_{n=1}^{N}\sum_{k=1}^{K} z_{nk} \norm{\textbf{x}_n - \bm{\mu}_k}^2 &\qquad (\text{K-Means Loss Function})
\end{flalign*}
With $\bm{\mu}_k \in \R^D$, $z_{nk} \in \{0,1\}$, $\sum_{k=1}^{K} z_{nk} = 1$ and where
\begin{alignat*}{2}
\textbf{z}_n &= [z_{n1},z_{n2},\dots,z_{nk}]^T &\qquad& \text{(indicator variable)}\\
\textbf{z} &= [\textbf{z}_1,\textbf{z}_2,\dots,\textbf{z}_N ]^T && \\
\bm{\mu} &= [\bm{\mu}_1,\bm{\mu}_2,\dots,\bm{\mu}_N]^T &&
\end{alignat*}
The indicator variable is an assignment of datapoint $n$ to a certain cluster, also called membership or latent variable (point $n$ belongs to cluster 3, $\textbf{z}_n = [0,0,1,0,\dots,0]^T$).

For fixed centres $\bm{\mu}_k$, the cost is minimized if we map each sample to its nearest centre, where we measure distance in terms of Euclidean distance.

% TODO The rest <.<


\end{document}
